File: frontend/services/agents.ts
/**
 * Multi-agent engine API service (Phase 3.5).
 *
 * POST /agents/run returns immediately with a run id; the pipeline executes
 * asynchronously on the backend and status/output are polled. Every contract
 * below mirrors the backend Pydantic schemas.
 */
import { z } from "zod";

import { http } from "./http-client";

// --- Schemas & types ----------------------------------------------------------

export const runStatusSchema = z.enum([
  "queued",
  "running",
  "completed",
  "failed",
  "cancelled",
]);
export type RunStatus = z.infer<typeof runStatusSchema>;

export const runModeSchema = z.enum(["llm", "deterministic"]);
export type RunMode = z.infer<typeof runModeSchema>;

export const stepStatusSchema = z.enum([
  "pending",
  "running",
  "completed",
  "failed",
  "skipped",
  "needs_revision",
]);
export type StepStatus = z.infer<typeof stepStatusSchema>;

export const agentDefinitionSchema = z.object({
  key: z.string(),
  name: z.string(),
  role: z.string(),
  description: z.string(),
  order: z.number(),
});
export type AgentDefinition = z.infer<typeof agentDefinitionSchema>;

export const runStatusResponseSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid().nullable(),
  status: runStatusSchema,
  mode: runModeSchema,
  current_step: z.string().nullable(),
  total_steps: z.number(),
  completed_steps: z.number(),
  progress: z.number().min(0).max(100),
  error: z.string().nullable(),
  started_at: z.string().nullable(),
  finished_at: z.string().nullable(),
  created_at: z.string(),
  // Review metadata (Phase 4.0)
  iteration: z.number().optional(),
  verdict: z.string().nullable(),
  overall_score: z.number().nullable(),
});
export type RunStatusResponse = z.infer<typeof runStatusResponseSchema>;

/** Agent artifact — a JSON object that always carries a `markdown` field. */
export const stepOutputSchema = z.record(z.string(), z.unknown());
export type StepOutput = z.infer<typeof stepOutputSchema>;

export const agentStepSchema = z.object({
  id: z.string().uuid(),
  agent: z.string(),
  status: stepStatusSchema,
  output: stepOutputSchema.nullable(),
  logs: z.array(z.string()).nullable(),
  duration_ms: z.number().nullable(),
  started_at: z.string().nullable(),
  finished_at: z.string().nullable(),
  // Execution metadata (Phase 4.0)
  iteration: z.number().optional(),
  input_artifacts: z.array(z.string()).nullable(),
  model_used: z.string().nullable(),
  token_usage: z.number().nullable(),
  feedback: z.array(z.string()).nullable(),
  error: z.string().nullable(),
});
export type AgentStep = z.infer<typeof agentStepSchema>;

export const runOutputSchema = z.object({
  run: runStatusResponseSchema,
  requirements: z.string().nullable(),
  steps: z.array(agentStepSchema),
});
export type RunOutput = z.infer<typeof runOutputSchema>;

export const runHistoryItemSchema = z.object({
  id: z.string().uuid(),
  status: runStatusSchema,
  mode: runModeSchema,
  requirements: z.string().nullable(),
  current_step: z.string().nullable(),
  progress: z.number().min(0).max(100),
  created_at: z.string(),
  finished_at: z.string().nullable(),
});
export type RunHistoryItem = z.infer<typeof runHistoryItemSchema>;

export const runHistoryPageSchema = z.object({
  items: z.array(runHistoryItemSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
});
export type RunHistoryPage = z.infer<typeof runHistoryPageSchema>;

export interface RunRequest {
  project_id?: string;
  requirements: string;
  preferred_stack?: string[];
  mode?: "auto" | "llm" | "deterministic";
}

export const runAcceptedSchema = z.object({
  run_id: z.string().uuid(),
  status: z.string(),
  mode: runModeSchema,
});
export type RunAccepted = z.infer<typeof runAcceptedSchema>;

// --- Helpers ------------------------------------------------------------------

export function stepMarkdown(step: AgentStep): string {
  const value = step.output?.markdown;
  return typeof value === "string" ? value : "";
}

// --- Service ------------------------------------------------------------------

export const agentsApi = {
  definitions: () => http.get("/api/v1/agents/definitions", z.array(agentDefinitionSchema)),

  run: (data: RunRequest) => http.post("/api/v1/agents/run", data, runAcceptedSchema),

  status: (id: string) => http.get(`/api/v1/agents/status/${id}`, runStatusResponseSchema),

  /** Request cancellation of a queued/running pipeline (Phase 3.6). */
  cancel: (id: string) =>
    http.post(`/api/v1/agents/cancel/${id}`, undefined, runStatusResponseSchema),

  /** Resume a failed run from the failed agent, reusing completed artifacts (Phase 4.0). */
  retry: (id: string) => http.post(`/api/v1/agents/retry/${id}`, undefined, runAcceptedSchema),

  output: (id: string) => http.get(`/api/v1/agents/output/${id}`, runOutputSchema),

  history: (params: { page?: number; page_size?: number; project_id?: string } = {}) => {
    const search = new URLSearchParams();
    if (params.page) {
      search.set("page", String(params.page));
    }
    search.set("page_size", String(params.page_size ?? 10));
    if (params.project_id) {
      search.set("project_id", params.project_id);
    }
    return http.get(`/api/v1/agents/history?${search.toString()}`, runHistoryPageSchema);
  },

  remove: (id: string) => http.delete(`/api/v1/agents/history/${id}`),
};

File: backend/app/api/v1/agents.py
"""Multi-agent engine endpoints (Phase 3.5)."""

from __future__ import annotations

import json
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Path, Query, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import SessionDep, SettingsDep, get_current_user, require_roles
from app.config.settings import Settings
from app.core.logging import get_logger
from app.models.user import Role, User
from app.schemas.agent import (
    AgentDefinitionResponse,
    AgentRunRequest,
    AgentStepResponse,
    RunAcceptedResponse,
    RunHistoryItem,
    RunHistoryResponse,
    RunOutputResponse,
    RunStatusResponse,
    agent_definitions,
)
from app.services import agent_service

logger = get_logger(__name__)

agents_router = APIRouter(prefix="/agents", tags=["agents"])

RunId = Annotated[uuid.UUID, Path(description="Agent run id")]


def _status(run) -> RunStatusResponse:
    return RunStatusResponse(
        id=run.id,
        project_id=run.project_id,
        status=run.status,
        mode=run.mode,
        current_step=run.current_step,
        total_steps=run.total_steps,
        completed_steps=run.completed_steps,
        progress=agent_service.progress(run),
        error=run.error,
        started_at=run.started_at,
        finished_at=run.finished_at,
        created_at=run.created_at,
        iteration=run.iteration or 1,
        verdict=run.verdict,
        overall_score=run.overall_score,
    )


def _step(step) -> AgentStepResponse:
    output: dict | None = None
    if step.output:
        try:
            output = json.loads(step.output)
        except json.JSONDecodeError:
            output = {"raw": step.output}
    logs: list[str] | None = None
    if step.logs:
        try:
            logs = json.loads(step.logs)
        except json.JSONDecodeError:
            logs = [step.logs]
    duration_ms = None
    if step.started_at and step.finished_at:
        duration_ms = round((step.finished_at - step.started_at).total_seconds() * 1000)
    return AgentStepResponse(
        id=step.id,
        agent=step.agent,
        status=step.status,
        output=output,
        logs=logs,
        duration_ms=duration_ms,
        started_at=step.started_at,
        finished_at=step.finished_at,
        iteration=step.iteration or 1,
        input_artifacts=step.input_artifacts,
        model_used=step.model_used,
        token_usage=step.token_usage,
        feedback=step.feedback,
        error=step.error,
    )


@agents_router.get(
    "/definitions",
    response_model=list[AgentDefinitionResponse],
    summary="Agent catalog",
    dependencies=[Depends(get_current_user)],
)
async def definitions() -> list[AgentDefinitionResponse]:
    return agent_definitions()


@agents_router.post(
    "/run",
    response_model=RunAcceptedResponse,
    status_code=202,
    summary="Start a multi-agent pipeline run",
    dependencies=[Depends(require_roles(Role.DEVELOPER, Role.ADMIN))],
)
async def run_pipeline(
    payload: AgentRunRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
    settings: Settings = SettingsDep,
) -> RunAcceptedResponse:
    resolved_mode = agent_service.resolve_mode(payload.mode, settings)
    run = await agent_service.create_run(
        db,
        current_user,
        project_id=payload.project_id,
        requirements=payload.requirements,
        preferred_stack=payload.preferred_stack,
        mode=resolved_mode,
    )
    agent_service.start_run(run.id)
    logger.info(
        "Agent run started {run_id} mode={mode}",
        run_id=run.id,
        mode=resolved_mode,
    )
    return RunAcceptedResponse(run_id=run.id, status=run.status, mode=run.mode)


@agents_router.post(
    "/retry/{run_id}",
    response_model=RunAcceptedResponse,
    status_code=202,
    summary="Retry a failed pipeline from the failed agent",
    dependencies=[Depends(require_roles(Role.DEVELOPER, Role.ADMIN))],
)
async def retry_run(
    run_id: RunId,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> RunAcceptedResponse:
    """Resume a failed run from the agent that failed, reusing completed
    artifacts (Phase 4.0). Returns the new run id."""
    run = await agent_service.retry_run(db, current_user, run_id)
    return RunAcceptedResponse(run_id=run.id, status=run.status, mode=run.mode)


@agents_router.post(
    "/cancel/{run_id}",
    response_model=RunStatusResponse,
    summary="Cancel a queued or running pipeline",
    dependencies=[Depends(require_roles(Role.DEVELOPER, Role.ADMIN))],
)
async def cancel_run(
    run_id: RunId,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> RunStatusResponse:
    run = await agent_service.cancel_run(db, current_user, run_id)
    return _status(run)


@agents_router.get("/status/{run_id}", response_model=RunStatusResponse, summary="Run status")
async def run_status(
    run_id: RunId,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> RunStatusResponse:
    run = await agent_service.get_run_for_user(db, current_user, run_id)
    return _status(run)


@agents_router.get("/output/{run_id}", response_model=RunOutputResponse, summary="Run output")
async def run_output(
    run_id: RunId,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> RunOutputResponse:
    run = await agent_service.get_run_for_user(db, current_user, run_id)
    steps = await agent_service.get_run_steps(db, run_id)
    return RunOutputResponse(
        run=_status(run),
        requirements=run.requirements,
        steps=[_step(step) for step in steps],
    )


def _preview(value: str | None, limit: int = 200) -> str | None:
    """Truncate long text for list payloads (requirements can reach 100k chars)."""
    if value is None or len(value) <= limit:
        return value
    return f"{value[: limit - 3]}..."


@agents_router.get("/history", response_model=RunHistoryResponse, summary="Run history")
async def run_history(
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 10,
    project_id: Annotated[uuid.UUID | None, Query(description="Filter by project")] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> RunHistoryResponse:
    runs, total = await agent_service.list_runs(
        db, current_user, page=page, page_size=page_size, project_id=project_id
    )
    return RunHistoryResponse(
        items=[
            RunHistoryItem(
                id=run.id,
                status=run.status,
                mode=run.mode,
                requirements=_preview(run.requirements),
                current_step=run.current_step,
                progress=agent_service.progress(run),
                created_at=run.created_at,
                finished_at=run.finished_at,
            )
            for run in runs
        ],
        total=total,
        page=page,
        page_size=page_size,
    )


@agents_router.delete(
    "/history/{run_id}",
    status_code=204,
    summary="Delete a run from history",
    dependencies=[Depends(require_roles(Role.DEVELOPER, Role.ADMIN))],
)
async def delete_history(
    run_id: RunId,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = SessionDep,
) -> Response:
    await agent_service.delete_run(db, current_user, run_id)
    return Response(status_code=204)
